kv/dist_sender.go

```diff
@@ -375,18 +379,19 @@ func (ds *DistSender) optimizeReplicaOrder(replicas replicaSlice) rpc.OrderingPo
 // until after the node has joined the gossip network and been allowed to initialize
 // its stores.
 func (ds *DistSender) getNodeDescriptor() *proto.NodeDescriptor {
-	if ds.nodeDescriptor != nil {
-		return ds.nodeDescriptor
+	if desc := atomic.LoadPointer(&ds.nodeDescriptor); desc != nil {
+		return (*proto.NodeDescriptor)(desc)
 	}
+
 	ownNodeID := ds.gossip.GetNodeID()
-	if nodeDesc, err := ds.gossip.GetInfo(
-		gossip.MakeNodeIDKey(ownNodeID)); err == nil && ownNodeID > 0 {
-		ds.nodeDescriptor = nodeDesc.(*proto.NodeDescriptor)
-	} else {
-		log.Infof("unable to determine this node's attributes for replica " +
-			"selection; node is most likely bootstrapping")
+
+	if nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(ownNodeID)); ownNodeID > 0 && err == nil {
+		atomic.StorePointer(&ds.nodeDescriptor, unsafe.Pointer(nodeDesc.(*proto.NodeDescriptor)))
```

Comment:
done (type assertion). I disagree about the lock because this operation is idempotent, but feel free to go for that after this lands. What would really be useful is avoiding this awkward dance around setting the NodeID late, but that's going to be more involved.